// vue-virtual-scroller/dist/composables/useTableColumnWidths.js
import { shallowRef as M, computed as b, toValue as s, watch as C, onBeforeUnmount as z, nextTick as T } from "vue";
function W() {
  return new Promise((e) => {
    requestAnimationFrame(() => e());
  });
}
function p(e) {
  if (e.borderBoxSize) {
    const t = Array.isArray(e.borderBoxSize) ? e.borderBoxSize[0] : e.borderBoxSize;
    if (t)
      return t.inlineSize;
  }
  return e.contentRect.width;
}
function x(e) {
  return (e.tHead ? [...e.tHead.rows] : []).at(-1) ?? null;
}
function S(e, t) {
  if (e.getAttribute("aria-hidden") === "true" || e.cells.length !== t)
    return !1;
  const n = getComputedStyle(e);
  return n.display === "none" || n.visibility === "hidden" ? !1 : e.getBoundingClientRect().width > 0;
}
function F(e) {
  if (e.getAttribute("aria-hidden") === "true" || e.cells.length === 0)
    return !1;
  const t = getComputedStyle(e);
  return t.display === "none" || t.visibility === "hidden" ? !1 : e.getBoundingClientRect().width > 0;
}
function H(e, t) {
  for (const n of [...e.tBodies])
    for (const i of [...n.rows])
      if (S(i, t))
        return i;
  return null;
}
function L(e) {
  const t = x(e);
  if (t != null && t.cells.length)
    return t.cells.length;
  for (const n of [...e.tBodies])
    for (const i of [...n.rows])
      if (F(i))
        return i.cells.length;
  return 0;
}
function v(e, t) {
  for (const [n, i] of [...e.cells].entries())
    t[n] = Math.max(t[n] ?? 0, Math.ceil(i.getBoundingClientRect().width));
}
function k(e) {
  const t = L(e);
  if (!t)
    return [];
  const n = Array.from({ length: t }, (r, l) => l).fill(0), i = x(e);
  if (i && i.cells.length === t && i.getBoundingClientRect().width > 0 && v(i, n), !H(e, t) && !i)
    return [];
  for (const r of [...e.tBodies])
    for (const l of [...r.rows])
      S(l, t) && v(l, n);
  return n.some((r) => r > 0) ? n : [];
}
function D(e) {
  const t = M([]), n = b(() => t.value), i = b(() => t.value.length > 0), y = b(() => t.value.length > 0 ? { tableLayout: "fixed" } : void 0);
  function r() {
    return s(e.disabled) || s(e.enabled ?? !0) === !1;
  }
  let l = !0, B = null, g = null, w = !1, m = !1, c = !1;
  const u = typeof ResizeObserver > "u" ? void 0 : new ResizeObserver((o) => {
    for (const h of o) {
      if (h.target !== B)
        continue;
      const f = p(h);
      f !== g && (g = f, d());
    }
  });
  function a() {
    t.value = [];
  }
  async function A() {
    if (m) {
      c = !0;
      return;
    }
    m = !0, a(), await T(), await W(), w = !1;
    const o = s(e.table);
    l && o && !r() && (t.value = k(o)), m = !1, c && (c = !1, d());
  }
  function d() {
    if (l) {
      if (r() || !s(e.table)) {
        a();
        return;
      }
      if (w) {
        c = !0;
        return;
      }
      w = !0, A();
    }
  }
  return C(
    () => [s(e.table), r()],
    ([o, h], f) => {
      const R = f == null ? void 0 : f[0];
      if (R && u && u.unobserve(R), B = o ?? null, g = (o == null ? void 0 : o.getBoundingClientRect().width) ?? null, !o || h) {
        a();
        return;
      }
      u == null || u.observe(o), d();
    },
    {
      immediate: !0,
      flush: "post"
    }
  ), C(
    () => s(e.dependencies) ?? null,
    () => {
      !r() && s(e.table) && d();
    },
    {
      deep: !0,
      flush: "post"
    }
  ), z(() => {
    l = !1, u == null || u.disconnect();
  }), {
    columnWidths: n,
    hasLockedWidths: i,
    tableStyle: y,
    scheduleMeasure: d,
    clear: a
  };
}
export {
  D as useTableColumnWidths
};
