// vue-virtual-scroller/dist/WindowScroller.vue_vue_type_script_setup_true_lang-Dpwklg_y.js
import { defineComponent as J, ref as S, toRef as K, computed as a, withDirectives as Q, openBlock as i, createElementBlock as d, normalizeClass as h, unref as t, renderSlot as n, createCommentVNode as s, createBlock as c, resolveDynamicComponent as b, normalizeStyle as f, withCtx as I, Fragment as X, renderList as Y, mergeProps as Z, createVNode as ee } from "vue";
import { useWindowScroller as te } from "./composables/useWindowScroller.js";
import { _ as le, R as ie, O as re } from "./scroller.css_vue_type_style_index_0_src_true_lang--7gJ3zh_.js";
import { r as se } from "./useRecycleScroller-CXk06epd.js";
const fe = /* @__PURE__ */ J({
  __name: "WindowScroller",
  props: {
    items: {},
    keyField: { default: "id" },
    direction: { default: "vertical" },
    listTag: { default: "div" },
    itemTag: { default: "div" },
    itemSize: { default: null },
    gridItems: { default: void 0 },
    itemSecondarySize: { default: void 0 },
    minItemSize: { default: null },
    sizeField: { default: "size" },
    typeField: { default: "type" },
    buffer: { default: 200 },
    shift: { type: Boolean, default: !1 },
    cache: { default: void 0 },
    prerender: { default: 0 },
    emitUpdate: { type: Boolean, default: !1 },
    disableTransform: { type: Boolean, default: !1 },
    flowMode: { type: Boolean, default: !1 },
    hiddenPosition: { default: void 0 },
    updateInterval: { default: 0 },
    enabled: { type: Boolean, default: !0 },
    listClass: { default: "" },
    itemClass: { default: "" }
  },
  emits: ["resize", "visible", "hidden", "update"],
  setup(k, { expose: T, emit: C }) {
    const e = k, o = C, F = re, m = S(), g = S(), z = S(), $ = K(e, "items"), V = a(() => ({
      items: $,
      el: m,
      before: g,
      after: z,
      keyField: e.keyField,
      direction: e.direction,
      itemSize: e.itemSize,
      gridItems: e.gridItems,
      itemSecondarySize: e.itemSecondarySize,
      minItemSize: e.minItemSize,
      sizeField: e.sizeField,
      typeField: e.typeField,
      buffer: e.buffer,
      shift: e.shift,
      cache: e.cache,
      prerender: e.prerender,
      emitUpdate: e.emitUpdate,
      disableTransform: e.disableTransform,
      flowMode: e.flowMode,
      hiddenPosition: e.hiddenPosition,
      updateInterval: e.updateInterval,
      enabled: e.enabled,
      onResize: () => o("resize"),
      onVisible: () => o("visible"),
      onHidden: () => o("hidden"),
      onUpdate: (l, w, r, v) => o("update", l, w, r, v)
    })), B = te(
      V
    ), {
      pool: M,
      totalSize: O,
      startSpacerSize: u,
      endSpacerSize: p,
      ready: _,
      scrollToItem: P,
      scrollToPosition: R,
      getScroll: x,
      findItemIndex: N,
      getItemOffset: U,
      getItemSize: W,
      getViewStyle: D,
      cacheSnapshot: H,
      restoreCache: E,
      updateVisibleItems: L,
      handleResize: j,
      handleVisibilityChange: q
    } = B, y = a(
      () => se({
        direction: e.direction,
        disableTransform: e.disableTransform,
        flowMode: e.flowMode,
        gridItems: e.gridItems
      }) === "flow"
    ), A = a(() => ({
      height: `${u.value}px`
    })), G = a(() => ({
      height: `${p.value}px`
    }));
    return T({
      el: m,
      startSpacerSize: u,
      endSpacerSize: p,
      scrollToItem: P,
      scrollToPosition: R,
      getScroll: x,
      findItemIndex: N,
      getItemOffset: U,
      getItemSize: W,
      cacheSnapshot: H,
      restoreCache: E,
      updateVisibleItems: L
    }), (l, w) => Q((i(), d("div", {
      ref_key: "el",
      ref: m,
      class: h(["vue-recycle-scroller vue-window-scroller", {
        "flow-mode": y.value,
        ready: t(_),
        [`direction-${e.direction}`]: !0
      }])
    }, [
      l.$slots.before ? (i(), d("div", {
        key: 0,
        ref_key: "before",
        ref: g,
        class: "vue-recycle-scroller__slot"
      }, [
        n(l.$slots, "before")
      ], 512)) : s("", !0),
      (i(), c(b(e.listTag), {
        style: f({ [e.direction === "vertical" ? "minHeight" : "minWidth"]: `${t(O)}px` }),
        class: h(["vue-recycle-scroller__item-wrapper", e.listClass])
      }, {
        default: I(() => [
          y.value && t(u) > 0 ? (i(), c(b(e.itemTag), {
            key: 0,
            "aria-hidden": "true",
            class: "vue-recycle-scroller__item-spacer",
            style: f(A.value)
          }, null, 8, ["style"])) : s("", !0),
          (i(!0), d(X, null, Y(t(M), (r) => (i(), c(le, {
            key: r.nr.id,
            view: r,
            "item-tag": e.itemTag,
            style: f(t(_) ? t(D)(r) : null),
            class: h(["vue-recycle-scroller__item-view", e.itemClass])
          }, {
            default: I((v) => [
              n(l.$slots, "default", Z({ ref_for: !0 }, v))
            ]),
            _: 2
          }, 1032, ["view", "item-tag", "style", "class"]))), 128)),
          y.value && t(p) > 0 ? (i(), c(b(e.itemTag), {
            key: 1,
            "aria-hidden": "true",
            class: "vue-recycle-scroller__item-spacer",
            style: f(G.value)
          }, null, 8, ["style"])) : s("", !0),
          e.items.length === 0 ? n(l.$slots, "empty", { key: 2 }) : s("", !0)
        ]),
        _: 3
      }, 8, ["style", "class"])),
      l.$slots.after ? (i(), d("div", {
        key: 1,
        ref_key: "after",
        ref: z,
        class: "vue-recycle-scroller__slot"
      }, [
        n(l.$slots, "after")
      ], 512)) : s("", !0),
      ee(ie, { onNotify: t(j) }, null, 8, ["onNotify"])
    ], 2)), [
      [t(F), t(q)]
    ]);
  }
});
export {
  fe as _
};
